From 627e17e1e2b1b09c95bd475ca9904f0b6c586f2a Mon Sep 17 00:00:00 2001
From: William Douglas <william.douglas@intel.com>
Date: Tue, 23 May 2023 11:23:15 -0700
Subject: [PATCH] clearlinux stateless configuration

Changes for the 1.4.21 release: fall back to /usr/share/defaults/mock
for bash completions and for the logging.ini file.

Changes for the 3.0 release: relocate DEFAULTMOCKCONFDIR setting to the
mockbuild/constants module, and add stateless support to new
simple_load_config function.

Changes for the 4.0 release: Rebase config.py, no code changes.

Ported-by: William Douglas <william.douglas@intel.com>
Signed-off-by: Simental Magana, Marcos <marcos.simental.magana@intel.com>
Signed-off-by: Patrick McCarty <patrick.mccarty@intel.com>
Signed-off-by: William Douglas <william.douglas@intel.com>
---
 docs/mock.1                |  9 +++++---
 etc/bash_completion.d/mock |  4 ++--
 etc/mock/clear.cfg         | 42 ++++++++++++++++++++++++++++++++++++++
 py/mock.py                 | 10 +++++----
 py/mockbuild/config.py     |  6 ++++--
 py/mockbuild/constants.py  |  1 +
 6 files changed, 61 insertions(+), 11 deletions(-)
 create mode 100644 etc/mock/clear.cfg

diff --git a/docs/mock.1 b/docs/mock.1
index 2b2f2c0..388093e 100644
--- a/docs/mock.1
+++ b/docs/mock.1
@@ -57,8 +57,9 @@ contents specified by a configuration file, then build any input SRPM(s) in
 that chroot.
 .LP
 The content of a chroot is specified by the configuration specified with the
-\fB\-r\fR option. The default configuration file is /etc/mock/default.cfg,
-which is usually a symlink to one of the installed configurations.
+\fB\-r\fR option. The default configuration file is
+/usr/share/defaults/mock/default.cfg, which is usually a symlink to one
+of the installed configurations.
 .LP
 There is a site\-wide configuration file, /etc/mock/site\-defaults.cfg, which can
 be used to specify site\-wide options. The shipped version of this file has no
@@ -447,7 +448,9 @@ don't create bootstrap chroot from container image
 
 .SH "FILES"
 .LP
-\fI/etc/mock/\fP \- default configuration directory
+\fI/usr/share/defaults/mock/\fP \- default configuration directory (used when /etc/mock/ is missing)
+.LP
+\fI/etc/mock/\fP \- system configuration directory
 .LP
 \fI/var/lib/mock\fP \- directory where chroots and results are created. You should not put there your input files.
 .SH "EXAMPLES"
diff --git a/etc/bash_completion.d/mock b/etc/bash_completion.d/mock
index 922a252..f472c03 100644
--- a/etc/bash_completion.d/mock
+++ b/etc/bash_completion.d/mock
@@ -2,7 +2,7 @@
 
 _mock_root()
 {
-    test $# -eq 0 && set -- /etc/mock "$HOME/.config/mock"
+    test $# -eq 0 && set -- /etc/mock "$HOME/.config/mock" /usr/share/defaults/mock
 
     local suggest=$(
       shopt -s nullglob
@@ -28,7 +28,7 @@ _mock()
     local cur prev words cword split
     _init_completion -s || return
 
-    local cfgdirs=( /etc/mock "$HOME/.config/mock" )
+    local cfgdirs=( /etc/mock "$HOME/.config/mock" /usr/share/defaults/mock )
     local count=0
     for word in "${words[@]}" ; do
         [[ $count -eq $cword ]] && break
diff --git a/etc/mock/clear.cfg b/etc/mock/clear.cfg
new file mode 100644
index 0000000..49e866b
--- /dev/null
+++ b/etc/mock/clear.cfg
@@ -0,0 +1,42 @@
+config_opts['root'] = 'clear'
+config_opts['target_arch'] = 'x86_64'
+config_opts['legal_host_arches'] = ('x86_64',)
+config_opts['chroot_setup_cmd'] = 'groupinstall build srpm-build'
+config_opts['dist'] = 'clear'  # only useful for --resultdir variable subst
+config_opts['extra_chroot_dirs'] = [ '/run/lock', ]
+config_opts['useradd'] = '/usr/sbin/useradd -m -u %(uid)s -g %(gid)s -d %(home)s  %(user)s'
+config_opts['plugin_conf']['ccache_enable'] = False
+config_opts['plugin_conf']['root_cache_opts']['compress_program'] = 'zstd -5'
+config_opts['plugin_conf']['root_cache_opts']['extension'] = '.zst'
+config_opts['releasever'] = 'clear'
+config_opts['package_manager'] = 'dnf'
+config_opts['chrootgroup'] = 'mockbuild'
+config_opts['use_bootstrap_container'] = False
+
+config_opts['dnf.conf'] = """
+[main]
+cachedir=/var/cache/dnf
+keepcache=1
+debuglevel=1
+reposdir=/dev/null
+logfile=/var/log/dnf.log
+retries=20
+obsoletes=1
+gpgcheck=0
+assumeyes=1
+syslog_ident=mock
+syslog_device=
+# repos
+[clearlinux]
+name=clearlinux
+baseurl=https://cdn.download.clearlinux.org/current/x86_64/os/
+gpgcheck=0
+cost=2000
+enabled=1
+[debuginfo]
+name=debuginfo
+baseurl=https://cdn.download.clearlinux.org/current/x86_64/debug/
+gpgcheck=0
+cost=2000
+enabled=1
+"""
diff --git a/py/mock.py b/py/mock.py
index aea83f2..75f50e5 100755
--- a/py/mock.py
+++ b/py/mock.py
@@ -62,7 +62,7 @@ import copy
 from functools import partial
 from mockbuild import config
 from mockbuild import util
-from mockbuild.constants import MOCKCONFDIR, PYTHONDIR, VERSION
+from mockbuild.constants import DEFAULTMOCKCONFDIR, MOCKCONFDIR, PYTHONDIR, VERSION
 from mockbuild.file_downloader import FileDownloader
 from mockbuild.mounts import BindMountPoint, FileSystemMountPoint
 
@@ -460,10 +460,10 @@ def setup_logging(config_path, config_opts, options):
 
     try:
         if not os.path.exists(log_ini):
-            if os.path.normpath('/etc/mock') != os.path.normpath(config_path):
+            if os.path.normpath(DEFAULTMOCKCONFDIR) != os.path.normpath(config_path):
                 log.warning("Could not find required logging config file: %s. Using default...",
                             log_ini)
-                log_ini = os.path.join("/etc/mock", config_opts["log_config_file"])
+                log_ini = os.path.join(DEFAULTMOCKCONFDIR, config_opts["log_config_file"])
                 if not os.path.exists(log_ini):
                     raise IOError("Could not find log config file %s" % log_ini)
             else:
@@ -684,7 +684,9 @@ def main():
         options.verbose = 0
 
     # config path -- can be overridden on cmdline
-    config_path = MOCKCONFDIR
+    config_path = DEFAULTMOCKCONFDIR
+    if os.path.exists(MOCKCONFDIR):
+        config_path = MOCKCONFDIR
     if options.configdir:
         config_path = options.configdir
 
diff --git a/py/mockbuild/config.py b/py/mockbuild/config.py
index f15f51c..5b2984b 100644
--- a/py/mockbuild/config.py
+++ b/py/mockbuild/config.py
@@ -20,7 +20,7 @@ import warnings
 from templated_dictionary import TemplatedDictionary
 from . import exception
 from . import text
-from .constants import MOCKCONFDIR, PKGPYTHONDIR, VERSION
+from .constants import DEFAULTMOCKCONFDIR, MOCKCONFDIR, PKGPYTHONDIR, VERSION
 from .file_util import is_in_dir
 from .trace_decorator import getLog, traceLog
 from .uid import getresuid, getresgid
@@ -799,7 +799,9 @@ def simple_load_config(name, config_path=None):
         break
 
     if config_path is None:
-        config_path = MOCKCONFDIR
+        config_path = DEFAULTMOCKCONFDIR
+        if os.path.exists(MOCKCONFDIR):
+            config_path = MOCKCONFDIR
     return load_config(config_path, name)
 
 
diff --git a/py/mockbuild/constants.py b/py/mockbuild/constants.py
index 6e3351e..91f752e 100644
--- a/py/mockbuild/constants.py
+++ b/py/mockbuild/constants.py
@@ -11,4 +11,5 @@ SYSCONFDIR = os.path.join(os.path.dirname(os.path.realpath(sys.argv[0])), "..",
 PYTHONDIR = os.path.dirname(os.path.realpath(sys.argv[0]))
 PKGPYTHONDIR = os.path.join(os.path.dirname(os.path.realpath(sys.argv[0])), "mockbuild")
 MOCKCONFDIR = os.path.join(SYSCONFDIR, "mock")
+DEFAULTMOCKCONFDIR = "/usr/share/defaults/mock"
 # end build system subs
-- 
2.40.1

